P2530R3 Hazard Pointers for C++26 (#6345)
cplusplus/draft: source/threads.tex

deadlock can occur.
\end{note}
\end{itemdescr}

\rSec2[saferecl.hp]{Hazard pointers}

\rSec3[saferecl.hp.general]{General}

\pnum
A hazard pointer is a single-writer multi-reader pointer
that can be owned by at most one thread at any time.
Only the owner of the hazard pointer can set its value,
while any number of threads may read its value.
The owner thread sets the value of a hazard pointer to point to an object
in order to indicate to concurrent threads---which
may delete such an object---that
the object is not yet safe to delete.

\pnum
A class type \tcode{T} is \defn{hazard-protectable}

> if it has exactly one base class of type \tcode{hazard_pointer_obj_base<T,D>}

Reviewer: ```suggestion
if it has exactly one base class of type \tcode{hazard_pointer_obj_base<T, D>}
```

Reviewer 2: Done.

for some \tcode{D},
that base is public and non-virtual, and

> it has no base classes of type \tcode{hazard_pointer_obj_base<T2,D2>}

Reviewer: ```suggestion
it has no base classes of type \tcode{hazard_pointer_obj_base<T2, D2>}
```

for any other combination \tcode{T2}, \tcode{D2}.
An object is \defn{hazard-protectable} if it is of hazard-protectable type.

\pnum
The time span between creation and destruction of a hazard pointer $h$

> is partitioned into a series of \defnadj{protection}{epochs};

Reviewer: ```suggestion
is partitioned into a series of \defnadjx{protection}{epochs}{epoch};
```

Reviewer 2: Done.

in each protection epoch,

> $h$ either is associated with a hazard-protectable object, or is unassociated.

Reviewer: The paper styles "associated with" and "unassociated" in italics.

Reviewer 2: That's true, but we don't always italicize local terms, e.g. see shared_ptr's "empty". I think this is fine.

Reviewer 2: On the other hand, "owns" below is a definition, so we might as well make this one one, too.

Upon creation, a hazard pointer is unassociated.
Changing the association (possibly to the same object)
initiates a new protection epoch and ends the preceding one.

\pnum
An object \tcode{x} of hazard-protectable type \tcode{T} is
\defn{retired} with a deleter of type \tcode{D}

> when the member function \tcode{hazard_pointer_obj_base<T,D>::retire}

Reviewer: ```suggestion
when the member function \tcode{hazard_pointer_obj_base<T, D>::retire}
```

Reviewer 2: Done.

\item
if $E$ began by an evaluation of \tcode{try_protect} with argument \tcode{src},
label its atomic load operation $L$.
If there exists an atomic modification $B$ on \tcode{src}
such that $L$ observes a modification that is modification-ordered before $B$, and
$B$ happens before \tcode{x} is retired,
the end of $E$ strongly happens before $A$.
\begin{note}
In typical use, a store to \tcode{src} sequenced before retiring \tcode{x}
will be such an atomic operation $B$.
\end{note}
\end{itemize}
\begin{note}
The latter two conditions convey the informal notion
that a protection epoch that began before retiring \tcode{x},
as implied either by the happens-before relation or
the coherence order of some source,
delays the reclamation of \tcode{x}.
\end{note}
\end{itemize}

\pnum
The number of possibly-reclaimable objects has an unspecified bound.
\begin{note}
The bound can be a function of the number of hazard pointers,
the number of threads that retire objects, and
the number of threads that use hazard pointers.
\end{note}
\begin{example}
The following example shows how hazard pointers allow updates to be carried out
in the presence of concurrent readers.

> The object of type hazard_pointer in \tcode{print_name}

Reviewer: ```suggestion
The object of type \tcode{hazard_pointer} in \tcode{print_name}
```

protects the object \tcode{*ptr} from being reclaimed by \tcode{ptr->retire}
until the end of the protection epoch.
\begin{codeblock}
struct Name : public hazard_pointer_obj_base<Name> { /* details */ };
atomic<Name*> name;
// called often and in parallel!
void print_name() {
  hazard_pointer h = make_hazard_pointer();
  Name* ptr = h.protect(name);          // Protection epoch starts
  // ... safe to access \tcode{*ptr}
}                                       // Protection epoch ends.

// called rarely, but possibly concurrently with \tcode{print_name}
void update_name(Name* new_name) {
  Name* ptr = name.exchange(new_name);
  ptr->retire();
}
\end{codeblock}
\end{example}

\rSec3[hazard.pointer.syn]{Header \tcode{<hazard_pointer>} synopsis}

> \begin{codeblock}

Reviewer: ```suggestion
\indexheader{hazard_pointer}%
\begin{codeblock}
```

namespace std {
  // \ref{saferecl.hp.base}, class template \tcode{hazard_pointer_obj_base}
  template<class T, class D = default_delete<T>> class hazard_pointer_obj_base;

>   // \ref{saferecl.hp.holder.general}, class \tcode{hazard_pointer}

Reviewer: ```suggestion
  // \ref{saferecl.hp.holder}, class \tcode{hazard_pointer}
```

\tcode{*this} is not empty.

\pnum
\effects
If \tcode{ptr} is a null pointer value, invokes \tcode{reset_protection()}.
Otherwise,
associates the hazard pointer owned by \tcode{*this} with \tcode{*ptr},
thereby ending the current protection epoch.

\pnum
\complexity
Constant.
\end{itemdescr}

\indexlibrarymember{reset_protection}{hazard_pointer}%
\begin{itemdecl}
void reset_protection(nullptr_t = nullptr) noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\expects
\tcode{*this} is not empty.

\pnum
\ensures
The hazard pointer owned by \tcode{*this} is unassociated.

\pnum
\complexity
Constant.
\end{itemdescr}

\indexlibrarymember{swap}{hazard_pointer}%
\begin{itemdecl}
void swap(hazard_pointer& other) noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Swaps the hazard pointer ownership of this object with that of \tcode{other}.
\begin{note}
The owned hazard pointers, if any, remain unchanged during the swap and
continue to be associated with the respective objects
that they were protecting before the swap, if any.
No protection epochs are ended or initiated.
\end{note}

\pnum
\complexity
Constant.
\end{itemdescr}

\rSec4[saferecl.hp.holder.nonmem]{Non-member functions}

> \indexlibrarymember{swap}{hazard_pointer}%

Reviewer: ```suggestion
\indexlibraryglobal{make_hazard_pointer}%
```
